src/stan_math_backend/Cpp.ml

```diff
@@ -0,0 +1,732 @@
+    | Var _ -> MethodCall (obj, name, templates, args)
+    | _ -> MethodCall (Parens obj, name, templates, args)
+
+  let to_var s = Var s
+
+  (** Turn an OCaml string into a quoted and escaped C++ string*)
+  let literal_string s = Literal ("\"" ^ Cpp_str.escaped s ^ "\"")
+
+  (** Equivalent to [std::vector<t>{e1,...,en}] *)
+  let std_vector_expr t es = InitializerExpr (Vector t, es)
+
+  let fun_call s es = FunCall (s, [], es)
+  let templated_fun_call s ts es = FunCall (s, ts, es)
+
+  (** Helper for [std::numeric_limits<double>::quiet_NaN()] *)
+  let quiet_NaN = fun_call "std::numeric_limits<double>::quiet_NaN" []
+
+  (** Helper for [std::numeric_limits<int>::min()] *)
+  let int_min = fun_call "std::numeric_limits<int>::min" []
+
+  (** Fold a list of expressions together with a C++ operator,
+      useful for e.g. adding a list of numbers. *)
+  let binop_list es ~f ~default : expr =
+    match es with
+    | [] -> default
+    | head :: rest -> List.fold ~init:head ~f:(fun a b -> Parens (f a b)) rest
+end
+
+module Expression_syntax = struct
+  (** Some operators to make streams and method calls look more
+      like the resultant C++ *)
+
+  include Exprs
+
+  (** A pun for the C++ [operator<<] *)
+  let ( << ) a b = StreamInsertion (a, b)
+
+  (** Method call: Call a no-argument method
+
+      E.g. [foo.bar()]
+  *)
+  let ( .@!() ) obj name = method_call obj name [] []
+
+  (** Method call: Call the named method with args
+
+      E.g. [foo.bar(A1,...An)]
+  *)
+  let ( .@?() ) obj (name, args) = method_call obj name [] args
+
+  (** Method call: Call the named method with template types and args
+
+      E.g. [foo.bar<T1,...,Tn>(A1,...An)]
+  *)
+  let ( .@<>() ) obj (name, templates, args) =
+    method_call obj name templates args
+
+  (** Static method call: Call the named method with no arguments.
+
+    E.g. [Foo::bar()]
+  *)
+  let ( |::! ) ty name = StaticMethodCall (ty, name, [], [])
+
+  (** Static method call: Call the named method with args
+
+    E.g. [Foo::bar(A1,...An)]
+  *)
+  let ( |::? ) ty (name, args) = StaticMethodCall (ty, name, [], args)
+
+  (** Static method call: Call the named method with template types and args
+
+    E.g. [Foo::bar<T1,...,Tn>(A1,...An)]
+  *)
+  let ( |::<> ) ty (name, templates, args) =
+    StaticMethodCall (ty, name, templates, args)
+
+  (** Pun for C++ [operator+(a,b)] *)
+  let ( + ) a b = BinOp (a, Add, b)
+
+  (** Pun for C++ [operator-(a,b)] *)
+  let ( - ) a b = BinOp (a, Subtract, b)
+
+  (** Pun for C++ [operator*(a,b)] *)
+  let ( * ) a b = BinOp (a, Multiply, b)
+end
+
+type init =
+  | Assignment of expr
+  | Construction of expr list
+  | InitalizerList of expr list
+  | Uninitialized
+[@@deriving sexp]
+
+type var_defn =
+  { static: bool [@default false]
+  ; constexpr: bool [@default false]
+  ; type_: type_
+  ; name: identifier
+  ; init: init [@default Uninitialized] }
+[@@deriving make, sexp]
+
+type stmt =
+  | Expression of expr
+  | VarDef of var_defn
```

Comment:
I'd rather have this be `obj_decl`. My thinking is that this is just the info that's used for the declaration of an object

Reply:
var is kind of overused because we have variables and the literal `var` autodiff type

Reply:
Do you mind `variable_defn`? It's nitpicky but there are plenty of types which aren't objects

Reply:
Let me read over this a bit more to see how it's used and we can circle back